### PR TITLE
on demand starting of mux mini-protocols & handler threads

### PR DESCRIPTION
**Pull request by dcoutts**
Work in progress on adjusting the mux to support starting mux mini-protocols on demand.

This is based on the ideas in #1349 but done on top of more recent mux refactoring & API changes, and taking a slightly different implementation & API approach. In particular we use fewer threads, by having the threads themselves start on demand, rather than having all threads start at the begning and blocking until they are needed.

### CODE REVIEW

File: network-mux/src/Network/Mux.hs [hunk 1]
```diff
@@ -228,27 +241,82 @@ muxStart tracer (MuxApplication ptcls) bearer = do
...
+         go (Set.delete ix idleResponders)
+
+        Right _ix ->
+          --TODO: it would be cleaner to do this check in the demuxer.
+          throwM $ MuxError MuxInitiatorOnly
+                      ("Received data on a responder channel but this mux "
+                       ++ "instance is initiator only")
+                   callStack
+
+    checkNonEmptyBuf (MiniProtocolDispatchInfo q _) = do
```
Comment (karknu) [inline network-mux/src/Network/Mux.hs]: A type signature for this would be nice.

File: network-mux/src/Network/Mux/Trace.hs [hunk 2]
```diff
@@ -49,9 +49,9 @@ data MuxErrorType = MuxUnknownMiniProtocol
                   | MuxIngressQueueOverRun
                   -- ^ thrown by 'demux' when violating 'maximumIngressQueue'
                   -- byte limit.
-                  | MuxControlProtocolError
-                  -- ^ thrown by 'muxControl' (mux control thread), when
-                  -- received a 'Muxcontrol' message on a mature 'MuxBearer'.
+                  | MuxInitiatorOnly
+                  -- ^ thrown when data arrives on a responder channel when the
+                  -- mux was set up as an 'InitiatorApp'.
```
Comment (karknu) [inline network-mux/src/Network/Mux/Trace.hs]: :+1: At least pre refactoring receiving a message on an initiator only bearer would have resulted in a `MuxUnknownMiniProtocol` error.

Comment (dcoutts) [reply network-mux/src/Network/Mux/Trace.hs]: I couldn't see how. The demuxer thread has a complete array of (protocol num, direction), with a queue for each one. It didn't know which queues had threads attached to them. The mux start would only start threads for the directions being used.

File: network-mux/src/Network/Mux.hs (hunk 1 above)
Comment (dcoutts) [reply network-mux/src/Network/Mux.hs]: done

File: network-mux/src/Network/Mux/Trace.hs (hunk 2 above)
Comment (karknu) [reply network-mux/src/Network/Mux/Trace.hs]: That may have been have been during a previous iteration of my InititatorAndResponder work.

**Pull request merged**